Ipfs (#1)
obsidiansystems/nix-rfcs: rfcs/0000-ipfs.md

This is a barrier to being a Nix user in areas of slower internet --- which includes the vast majority of the world's population at this time.
This is also a barrier to users running their own caches.

Content-addressing opens up a *huge* design space of solutions to get around such problems.
IPFS explores many of those solutions.

## Source distribution and archival

A goal of the Nix ecosystem is to package software in a way that never bitrots.
Getting in the way of that, however, is the fact source code frequently goes off-line.
The Software Heritage archive is the best in the world, and a natural partner in this effort.

Unfortunately, as https://www.tweag.io/blog/2020-06-18-software-heritage/ describes at the end, a major challenge is the way nix content-addresses software.
First of all, Nix hashes sources in bespoke ways that no other project will adopt.
Second of all, tarballs instead of the underlying files leaking non-normative details (compression, odd perms, etc.).

We should natively support git file hashing, which Git repos and Software Heritage both support.
This will completely obliterate these issues.

> IPFS also supports git hashing, and so we also provide a good way for people and intuitions to "pin" the sources they need, especially if those sources include private ones SWH won't have.

Author: people and institutions

Reviewer: ```suggestion
IPFS also supports git hashing, and so we also provide a good way for people and institutions to "pin" the sources they need, especially if those sources include private ones SWH won't have.
```

> Finally, per [Obsidian's bridging work](https://github.com/obsidiansystems/go-ipfs-swh-plugin),

Author: finish this thought

Reviewer: ```suggestion
Finally, per [Obsidian's bridging work](https://github.com/obsidiansystems/go-ipfs-swh-plugin), we have a 3 way integration between IPFS, SWH, and Nix. 
Data can be directly downloaded from SWH via HTTPS, or indirectly via IPFS, which can act as a CDN to not put as much load on SWH's servers.

Overall, are building out a uniform way to work with source code, regardless of its origins or the exact tools involved.
```

## Not just IPFS

Many of the IPFS-specific logic could in fact live in a plugin if this is desired.
However, we still need to adjust core abstractions of Nix store layer (as described below) to interface with IPFS in the best possible way.
Those same adjustments would allow Nix to work better with *any* content-addressing system, so alternatives networks/projects to IPFS can also be just as easily experimented with.

> As always with my work, the manta (from Scheme) to follow is

Author: mantra

Reviewer: ```suggestion
As always with my work, the mantra (from Scheme) to follow is
```

> *x* should be designed not by piling feature on top of feature, but by removing the weaknesses and restrictions that make additional features appear necessary.

A ton of misc features have been added to Nix since 2.0, and we are very careful to not increase total ad-hoc complexity more than necessary.

## Build adoption through seamless interop

This last argument is more strategic than technical.

A lot of people in this community would like to see Nix be used more widely, but as much as we all wish otherwise, the fact remains that there is some tension between making nix *better* and making it *more accessible*.

Nix is very foreign from the "bad conventional" way things are done, and making Nix better can sometimes involve making it even more foreign.
We don't want to steepen the learning curve or make it "seem more weird".

On the other hand, making Nix more accessible by making it more like tools users are already use-to can obscure or chip-away at Nix's benefits.

> We don't want to "pander" in ways that will make Nix faddish but ultimately undermine it's popular over the long haul (see Docker the company's woes).

Author: popularity

Reviewer: ```suggestion
We don't want to "pander" in ways that will make Nix faddish but ultimately undermine it's popularity over the long haul (see Docker the company's woes).
```

Like with "IPFS Narinfo", this format is also very easy to understand with nix-agnostic native IPFS tools.
This is because, once again, the reference graph is made native to IPFS not done indirectly with store path strings which must be looked up.

An interesting corollary to note:
Content addressing today is "shallow", in that references are arbitrary store paths.
With this form on content addressing, references are instead CIDs (native IPFS references) to other obligatorily content-addressed data.
This means the content addressing is "deep", such that any such content-addressed store object always has a content-addressed closure.
At the cost of interop with existing derivation outputs, this make such data easier to manage because there are fewer trust issues and degrees of freedom in general for something to go wrong.

## IPLD Derivations

- **Purpose**: Build plan distribution
- **Depends on**: Wrapped git objects with references,
                  IPFS as substitutor

Natively represent derivations in IPFS, again with the same benefits of leverage the native graph representations.

> This is a culmination of all the futures so far.

Author: features

Reviewer: ```suggestion
This is a culmination of all the features so far.
```

The derivations must be CA derivations (floating or fixed).
They must also produce wrapped git objects with references, though they can also depend on regular unwrapped git file hashed store objects.

The derivations and their outputs are thus all fully IPFS native, leveraging the IPFS graph and trust vs plain old data separation for the high standard of interoperability.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

We encourage anyone interested to check our tutorial in https://github.com/obsidiansystems/ipfs-nix-guide/ which demonstrates the above functionality.
Note at the time of writing this guide uses our original 2020 fork of Nix.

# Drawbacks
[drawbacks]: #drawbacks

The main cost is more complexity to the store layer.
For a few reason we think this is not so bad.

Most importantly is the division of the work into a dependency graph of steps.
This allows us to slowly try IPFS out and not commit to more change than we want to up front.

> Even if we do end up adopting everything though, we thing for the following two reasons the complexity can still be kept manageable:

Author: think

Reviewer: ```suggestion
Even if we do end up adopting everything though, we think for the following two reasons the complexity can still be kept manageable:
```

2. Per the abstract vs concrete model of the nix store in https://github.com/NixOS/nix/pull/6877 , everything we are doing is simply flushing out alternative interpretations of the abstract model.
   This is the sense in which we are "removing the weaknesses and restrictions that make additional features appear necessary" per the Scheme mantra cited above:

>    Instead of extending the model wit new features, we are reflaxing concrete model assumptions (e.g. references are always opaque store paths) while keeping the abstract model the same.

Author: with

Author: reflaxing?

Reviewer: ```suggestion
   Instead of extending the model with new features, we are relaxing concrete model assumptions (e.g. references are always opaque store paths) while keeping the abstract model the same.
```
